r/src/altrep.cpp

```diff
@@ -366,6 +388,277 @@ struct AltrepVectorPrimitive : public AltrepVectorBase<AltrepVectorPrimitive<sex
+    }
+
+    return nullptr;
+  }
+
+  static void* Dataptr(SEXP alt, Rboolean writeable) { return DATAPTR(Materialize(alt)); }
+
+  static SEXP Duplicate(SEXP alt, Rboolean /* deep */) {
+    // the representation integer vector
+    SEXP dup = PROTECT(Rf_lazy_duplicate(Materialize(alt)));
+
+    // additional attributes from the altrep
+    SEXP atts = PROTECT(Rf_duplicate(ATTRIB(alt)));
+    SET_ATTRIB(dup, atts);
+
+    UNPROTECT(2);
+    return dup;
+  }
+
+  // The value at position i
+  static int Elt(SEXP alt, R_xlen_t i) {
+    if (Base::IsMaterialized(alt)) {
+      return INTEGER_ELT(Representation(alt), i);
+    }
+
+    int out;
+    Get_region(alt, i, 1, &out);
+    return out;
+  }
+
+  static R_xlen_t Get_region(SEXP alt, R_xlen_t start, R_xlen_t n, int* buf) {
+    // If we have data2, we can just copy the region into buf
+    // using the standard Get_region for this R type
+    if (Base::IsMaterialized(alt)) {
+      return Standard_Get_region<int>(Representation(alt), start, n, buf);
+    }
+
+    auto chunked_array = GetChunkedArray(alt);
+
+    // get out if there is nothing to do
+    auto chunked_array_size = chunked_array->length();
+    if (start >= chunked_array_size) return 0;
+
+    auto slice = GetChunkedArray(alt)->Slice(start, n);
+
+    if (WasUnified(alt)) {
+      int j = 0;
+
+      // find out which is the first chunk of the chunk array
+      // that is present in the slice, because the main loop
+      // needs to refer to the correct transpose buffers
+      int64_t k = 0;
+      for (; j < chunked_array->num_chunks(); j++) {
+        auto nj = chunked_array->chunk(j)->length();
+        if (k + nj > start) {
+          break;
+        }
+
+        k += nj;
+      }
+
+      int* out = buf;
+      for (const auto& array : slice->chunks()) {
+        const auto& indices =
+            internal::checked_cast<const DictionaryArray&>(*array).indices();
+
+        // using the transpose data for this chunk
+        const auto* transpose_data =
+            reinterpret_cast<const int32_t*>(GetArrayTransposed(alt, j)->data());
+        auto transpose = [transpose_data](int x) { return transpose_data[x]; };
+
+        GetRegionDispatch(array, indices, transpose, out);
+
+        out += array->length();
+        j++;
+      }
+
+    } else {
+      // simpler case, identity transpose
+      auto transpose = [](int x) { return x; };
+
+      int* out = buf;
+      for (const auto& array : slice->chunks()) {
+        const auto& indices =
+            internal::checked_cast<const DictionaryArray&>(*array).indices();
+
+        GetRegionDispatch(array, indices, transpose, out);
+
+        out += array->length();
+      }
+    }
+
+    return slice->length();
+  }
+
+  template <typename Transpose>
+  static void GetRegionDispatch(const std::shared_ptr<Array>& array,
+                                const std::shared_ptr<Array>& indices,
+                                Transpose transpose, int* out) {
```

Comment:
Pass `Transpose&& transpose` and pass `std::forward<Transpose>(transpose)` below.